docs/hypermail.1
.TH HYPERMAIL 1 "June 24, 2000"
.SH NAME
hypermail \- convert mail archives in UNIX box format to HTML pages
.SH SYNOPSIS
.BR "hypermail " [ \-AgiMptTuvVxX1 ]
.RB [ \-m
.IR "mailbox" ]
.RB [ \-d
.IR "directory" ]
.RB [ \-l
.IR "label" ]
.RB [ \-L
.IR "language" ]
.RB [ \-a
.IR "URL" ]
.RB [ \-b
.IR "URL" ]
.RB [ \-c
.IR "file" ]
.RB [ \-n
.IR "listaddress" ]
.RB [ \-o
.IR "keyword=value" ]
.RB [ \-s
.IR "htmlsuffix" ]
.RB [ \-0
.IR "number" ]
.RB [ "mailbox" ]
.SH DESCRIPTION
.B hypermail
is a program that takes a file of mail messages in UNIX mailbox format and generates a set of cross-referenced HTML documents.  Each file that is created represents a separate message in the mail archive and contains links to other articles, so that the entire archive can be browsed in a number of ways by following links.  Archives generated by Hypermail can be incrementally updated, and Hypermail is set by default to only update archives when changes are detected.
.LP
Each HTML file that is generated for a message contains (where applicable): the subject of the article, the name and email address of the sender, the date the article was sent, links to the next and previous messages in the archive, a link to the message the article is in reply to, and a link to the message next in the current thread.
.LP
In addition, Hypermail will convert references in each message to email addresses and URLs to hyperlinks so they can be selected.  Email addresses will be converted to 
.B mailto:
URLs, or links to a CGI mail program. 
.LP
To complement each set of HTML messages, four
.I index files
are created which sort the articles by date received, thread, subject, and author.  Each entry in these index files are links to the individual articles and provide a bird's-eye view of every archived message:
.TP
.B date.html
The index of articles sorted by the date they were received by the mail daemon.
.TP
.B thread.html
The index of articles sorted by thread first, then the date they were received.
.TP
.B subject.html
The index of articles sorted by subject.  Any
.B "Re:"
prefixes in front of subjects will have been stripped out.
.TP
.B author.html
is the index of articles sorted by the first word of the author's name.  If the author's name can't be determined, their email address will be substituted.
.LP
One of the index files will be called
.B index.html
and is the default index that users can go to when entering the archive.  In the specified directory, articles will be read out in the order that they were read from a mailbox or standard input.  Filenames start at zero and increase
in this fashion:
.B 0000.html, 0001.html, 0002.html,
etc.
.SH OPTIONS
.TP
.BI \-a " URL"
This option includes a link labelled
.B "Other mail archives"
in the index pages to the specified URL.  This way users who are looking at the Hypermail archive have the opportunity to go to pointers to other mail archives.  By default, this is a link to the parent directory which holds the archive files.
.TP
.BI \-A
Use this to maintain a parallel mbox archive. The file
name defaults to mbox in the directory specified by -d or dir.
.TP
.BI \-b " URL"
This option includes a link labelled
.B "About this archive"
in the index pages to the specified URL.  This way users who are looking at the Hypermail archive have the opportunity to go to information about the archive.
.TP
.BI \-c " file"
This option specifies a configuration file to read settings from.  By default, Hypermail will look for a file called
.B .hmrc
in the user's home directory.
.TP
.BI \-d " directory"
Specifies the directory to put the HTML files and index files that are created.  If the directory doesn't exist, a new one will be created with the name that is specified.  If the
.B \-d
option isn't used, Hypermail will look for a directory with the same name as the input mailbox or will create one if needed. 
.TP
.B \-g
Use this to use gdbm to implement a header cache.
This will speed up hypermail, especially if your filesystem is slow.
It will probably not provide any speedup if you use the linkquotes option.
.TP
.B \-i
Reads in articles from standard input.
.TP
.BI \-l " label"
This option tells Hypermail what to call the archive \- the name that is specified will be in the title of the index pages so users know what sort of messages are being archived.
.TP
.BI \-L " language"
 This is a two-letter string specifying the default
language  to  use, or a longer string specifying a language
and locale. Set this the value of the language
table you wish  to  use  when  running  and  generating
archives.  See also the iso2022jp and eurodate config file options.
 Current supported languages, with their default locales:
 de (de_DE) - German
 en (en_US) - English
 es (es_ES) - Spanish
 fi (fi_FI) - Finnish
 fr (fr_FR) - French
 el (el) - Greek
 gr (el_GR) - Greek
 is (is_IS) - Icelandic
 no (no_NO) - Norwegian
 pl (pl_PL) - Polish
 pt (pt_BR) - Brazilian Portuguese
 ru (ru_RU) - Russian
 sv (sv_SE) - Swedish
.TP
.BI \-m " mailbox"
Specifies the mailbox to read articles in from.  By default, Hypermail will look for a file called
.B mbox.
.TP
.B \-M
 This option allows you to use metadata to store the content type
of a MIME attachments and, later on,  when a user browses the
attachment, send back this information in the HTTP Content-Type
header. When used, the Content-Type header of a MIME
attachment will be stored in a metadata file. 
.TP
.BI \-n " submission-address"
This is the list\'s submission address.  In this manner people will be able 
to submit new messages to the list the hypermail archive serves.
.TP
.BI \-o " keyword=value"
This is a means of setting any variable that can be specified in a config
file.
.TP
.B \-y
This tells Hypermail to do a
.B dry\-run
that won't generate any archive; however it will output
the file names and Message-IDs associated with the archive. It can help you visualize the
issues you may have when upgrading from an older Hypermail archive to a more recent one.
.TP
.B \-p
This shows a progress report as Hypermail reads in and writes out messages \- the number of files that Hypermail is reading and writing and the file names of the directory and files created are shown.
.TP
.B \-s htmlsuffix
Use this to specify the html file suffix to be used when Hypermail
generates the html files. This is dependent on local needs. Do not 
put a '.' in the value. It would result in "file..html", probably
not what you want.
.TP
.B \-u
This option tells Hypermail to add message(s) to the end of the existing HTML file archive and integrate them into it by links and cross-references. All archive index files will be regenerated to include the new message.
 Hypermail used to require that you only send it one message at a time when
using the -u option, but it should now work reasonably when
given mailboxes containing multiple messages.
 When using the -u option, don't send any messages that
Hypermail has already processed. If you want Hypermail to recognize that
some messages are old messages that shouldn't be added to the archive again,
send it a mailbox with a complete set of messages and avoid the -u option. 
.TP
.B \-v
This shows a the variables and their values that Hypermail will use when.
.TP
.B \-V
This shows the version information for the executing Hypermail.
Once displayed, Hypermail exits without doing any processing.
.TP
.B \-x
This tells Hypermail to explicitly overwrite any previous HTML files that may exist.  Use this option only when it is desirable to completely rewrite the entire archive.
.TP
.B \-X
Use this to let hypermail write an XML archive overview file in each
directory. The filename is archive_overview.haof.
.TP
.B \-0 number
 This is a message number that should be deleted from the
html archive. The mbox is not changed.
 See the delete_level config file option for more info about
what happens to the message.
.TP
.B \-1
 Use this to specify there is only one message in the input.
.LP
.SS
GENERAL EXECUTION NOTES
.LP
.B Note:
No matter what options are specified, the index files are always rewritten.  The date when Hypermail was last run is included in index pages, so it's easy to tell when the archive was last updated.
.LP
.B Note:
The
.B \-i
and
.B \-m
options cannot be used together.  Only archives in UNIX mailbox format can be read in - mailboxes of this kind are usually appended RFC2822-compliant articles separated by lines such as "\\nFrom person@site Mon Jan 10 12:34:56 1994".
.LP
.B Note:
If the mailbox that is being read from is an archive that new messages are always being added to, don't use the
.B \-u
option.  Hypermail will then read in all the messages given it but will only write new messages that have been appended to the mailbox.
.LP
.SS
CONFIGURATION OPTIONS
.LP
The following settings can be read in as environment variables or from the specified configuration file.  Environment settings are in uppercase.  For instance, in the C shell, variables can be set as:
.LP
.PD 0
.RS
.B setenv HM_MBOX "/home/john/my_mailbox"
.LP
.B setenv HM_FILEMODE 0600
.PD
.RE
.LP
In the configuration file, blank lines and lines beginning with a hash mark
.RB ( # )
are ignored.  Variables must be in lowercase and separated by values with an equals
.RB ( = )
sign, such as:
.LP
.PD 0
.RS
.B mbox = "/home/john/my_mailbox"
.LP
.B filemode = 0600
.PD
.RE
.LP
Settings are read in this order: from the program's hard-wired internal defaults, from environment variables, from the configuration file, from command-line options.
.LP
See 
.B hmrc.4
for more information on configuration file usage.
.LP
Below is a partial list of variables that Hypermail understands. A full list is available in the file hmrc.html, or you can also look in setup.c. Boolean numbers can have the value of
.B 0
or
.B 1.
.TP
.B HM_CONFIGFILE "filename"
This is the default configuration file to read settings in from.  This can only be specified as an environment variable.  If the first character is "~", Hypermail will look for the file under the current user's home directory.
.TP
.B HM_MBOX "filename"
This is the default mailbox to read messages in from.  Define this with a value of
.B "NONE"
to read from standard input as the default.
.TP
.B HM_ARCHIVES "URL"
This will create a link in the archived index pages labelled
.I "Other mail archives"
to the specified URL.  Define as
.B "NONE"
to omit such a link.
.TP
.B HM_ABOUT "URL"
This will create a link in the archived index pages labelled
.I "About this archive"
to the specified URL.  Define as
.B "NONE"
to omit such a link.
.TP
.B HM_REVERSE boolean_number
Defining this variable as
.B 1
will reverse-sort the article entries in the date and thread index files by the date they were received.  That is, the most recent messages will appear at the top of the index rather than the other way around.
.TP
.B HM_SHOWHEADERS boolean_number
Define this as
.B 1
 to show the article header lines in the archived HTML files.  These lines 
typically include the
.BR "To:" , " From: " and " Subject:"
information found in most email messages.
.TP
.B HM_SHOWHTML 0, 1, or 2
Define as
.B 1
to show the articles in a proportionally-spaced
font rather than a fixed-width (monospace) font. Setting this 
option to 1 also tells Hypermail to attempt to italicize quoted 
passages in articles.
Define as
.B 2
for more complex conversion to html
similar to that in txt2html.pl.

src/hypermail.c
/* 
** Copyright (C) 1994, 1995 Enterprise Integration Technologies Corp.
**         VeriFone Inc./Hewlett-Packard. All Rights Reserved.
** Kevin Hughes, kev@kevcom.com 3/11/94
** Kent Landfield, kent@landfield.com 4/6/97
** Hypermail Project 1998-2023
** 
** This program and library is free software; you can redistribute it and/or 
** modify it under the terms of the GNU (Library) General Public License 
** as published by the Free Software Foundation; either version 3
** of the License, or any later version. 
** 
** This program is distributed in the hope that it will be useful, 
** but WITHOUT ANY WARRANTY; without even the implied warranty of 
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
** GNU (Library) General Public License for more details. 
** 
** You should have received a copy of the GNU (Library) General Public License
** along with this program; if not, write to the Free Software 
** Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA 
*/

#define MAIN_FILE 1

#include "hypermail.h"

#ifdef __LCC__
#include "../lcc/defaults.h"
#else
#include "defaults.h"
#endif

#include "setup.h"
#include "parse.h"
#include "print.h"
#include "finelink.h"
#include "search.h"
#include "struct.h"

#ifdef HAVE_LOCALE_H
#include <locale.h>
#endif

/* #include <stdlib.h> */

extern int getopt(int, char *const *, const char *);
extern char *optarg;
extern int optind;
extern int opterr;

/*
** Prints user startup errors stops. 
*/
void cmderr(char *errorstr)
{
    fprintf(stderr, "\n%s: %s: %s\n", lang[MSG_ERROR], PROGNAME, errorstr);
    usage();
    unlock_archive();
    exit(-1);
}

/*
** Prints a program error string and stops. 
*/

void progerr(char *errorstr)
{
    fprintf(stderr, "%s: %s\n", PROGNAME, errorstr);
    unlock_archive();
    exit(-1);
}

static char *expand_contents(char *variable)
{
    char *return_value = NULL;

    /*
     * Potentially read the header and footer files. Check to make
     * sure the value is a valid file or else assure the original
     * value is there.
     */
    if (variable) {
	/*
	 * Check if valid filename. If not then just return the
	 * variable listed. 
	 * NOTE !! This really should be an error message!
	 */
	if (isfile(variable)) {
	    /* 
	     * Now try to assign the contents of the file to
	     * the return value. If it is able to read the file
	     * then free the existing storage and return.
	     * If not then just return the existing value.
	     * NOTE !! This really should be an error message!
	     */
	    return_value = getfilecontents(variable);
	}
    }
    return return_value;
}

char *setindex(char *dfltindex, char *indextype, char *suffix)
{
    char *p, *rp;

    p = (!strcmp(dfltindex, indextype)) ? INDEXNAME : indextype;
    rp = (char *)emalloc(strlen(p) + strlen(suffix) + 2);
    sprintf(rp, "%s.%s", p, suffix);

    return (rp);
}


/* Print out the version number and die. */

void version(void)
{
    printf("%s: %s: %s\n", PROGNAME, lang[MSG_VERSION], VERSION);
    exit(0);
}

/*
** Prints the usage.
*/

void usage(void)
{
    struct language_entry *lte;

    printf("%s: %s [options]\n", lang[MSG_USAGE], PROGNAME);

    printf("%s:\n", lang[MSG_OPTIONS]);
    printf("  -a URL        : %s\n", lang[MSG_OPTION_A]);
    printf("  -A            : %s\n", lang[MSG_OPTION_ARCHIVE]);
    printf("  -b URL        : %s\n", lang[MSG_OPTION_B]);
    printf("  -c file       : %s\n", lang[MSG_OPTION_C]);
    printf("  -d dir        : %s\n", lang[MSG_OPTION_D]);
#ifdef GDBM
    printf("  -g            : %s\n", lang[MSG_OPTION_G]);
#else
    printf("  [ -g            : %s ]\n", lang[MSG_OPTION_G_NOT_BUILD_IN]);
#endif
    printf("  -i            : %s\n", lang[MSG_OPTION_I]);
    printf("  -l label      : %s\n", lang[MSG_OPTION_L]);
    printf("  -m mbox       : %s\n", lang[MSG_OPTION_M]);
    printf("  -M            : %s\n", "Use metadata");
    printf("  -n listaddr   : %s\n", lang[MSG_OPTION_N]);
    printf("  -o keyword=val: Set config item\n");
    printf("  -p            : %s\n", lang[MSG_OPTION_P]);
    printf("  -s htmlsuffix : %s\n", "HTML file suffix (.html, .htm, ..)");
    printf("  -u            : %s\n", lang[MSG_OPTION_U]);
    printf("  -v            : %s\n", lang[MSG_OPTION_VERBOSE]);
    printf("  -V            : %s\n", lang[MSG_OPTION_VERSION]);
    printf("  -x            : %s\n", lang[MSG_OPTION_X]);
    printf("  -X            : %s\n", lang[MSG_OPTION_XML]);
    printf("  -1            : %s\n", lang[MSG_OPTION_1]);
    printf("  -L lang       : %s (", lang[MSG_OPTION_LANG]);

    /* Print out languages supported */
    lte = &ltable[0];
    while (lte->langcode != NULL) {
	printf("%s ", lte->langcode);
	lte++;
    }
    printf(")\n");
    printf("%s         : %s\n", lang[MSG_VERSION], VERSION);
    printf("%s            : %s\n\n", lang[MSG_DOCS], HMURL);
    exit(1);
}

int main(int argc, char **argv)
{
    int i, use_stdin, use_mbox;
    char *configfile = NULL;
    char **tlang, *locale_code;
    int cmd_show_variables;
    int print_usage;

    int amount_old = 0;		/* number of old mails */
    int amount_new = 0;		/* number of new mails */

#ifdef HAVE_LOCALE_H
    setlocale(LC_ALL, "");
#endif

    lockfile[0] = '\0';
    use_stdin = 0;
    print_usage = 0;
    use_mbox = 0;

    firstdatenum = lastdatenum = 0;

    configfile = strsav(CONFIGFILE);

    cmd_show_variables = 0;

    opterr = 0;

#define GETOPT_OPTSTRING ("a:Ab:c:d:gil:L:m:n:o:ps:tTuvVxX0:1M?")

    /* get pre config options here */
	while ((i = getopt(argc, argv, GETOPT_OPTSTRING)) != -1) {
		switch ((char)i) {
	case 'c':
	    configfile = strreplace(configfile, optarg);
	    break;
	case 'v':
	    cmd_show_variables = TRUE;
	    break;
	case 'V':
	    version();
		 /*NOTREACHED*/ case 'a':
	case 'A':
	case 'b':
	case 'd':
	case 'g':
	case 'i':
	case 'l':
	case 'L':
	case 'm':
	case 'n':
	case 'o':
	case 'p':
	case 's':
	case 't':
	case 'u':
	case 'x':
	case 'X':
	case '0':
	case '1':
	case 'M':
	    break;
	case '?':
	default:
	    /* 
	     * Because we need to setup the language support, 
	     * printing of the usage message must be deferred 
	     * until the proper language is determined.
	     */
	    print_usage = 1;
	    break;
	}
    }

    /* 
     * ...then read the configuration file.
     */

    readconfigs(configfile, cmd_show_variables);

    /* reset the getopt() index variable */
    optind = 1;

    /* now get the post-config options! */

	while ((i = getopt(argc, argv, GETOPT_OPTSTRING)) != -1) {
		switch ((char)i) {
	case 'A':
	    set_append = 1;
	    break;
	case 'a':
	    set_archives = strreplace(set_archives, optarg);
	    break;
	case 'b':
	    set_about = strreplace(set_about, optarg);
	    break;
	case 'c':
	    /* config file from pre-config options */
	    break;
	case 'd':
	    set_dir = strreplace(set_dir, optarg);
	    break;
	case 'g':
	    set_usegdbm = 1;
	    break;
	case 'i':
	    use_stdin = TRUE;
	    break;
	case 'l':
	    set_label = strreplace(set_label, optarg);
	    break;
	case 'L':
	    set_language = strreplace(set_language, optarg);
	    break;
	case 'm':
	    set_mbox = strreplace(set_mbox, optarg);
	    break;
	case 'n':
	    set_hmail = strreplace(set_hmail, optarg);
	    break;
	case 'o':
	    ConfigAddItem(optarg);
	    break;
	case 'p':
	    set_showprogress = TRUE;
	    break;
	case 's':
	    set_htmlsuffix = strreplace(set_htmlsuffix, optarg);
	    break;
	case 't':
	    set_usetable = TRUE;
	    break;
            /* removed in 2.2.25 */
            /*
	case 'T':
	    set_indextable = TRUE;
	    break;
            */
	case 'u':
	    set_increment = TRUE;
	    break;
	case 'v':
	    cmd_show_variables = TRUE;
	    break;
	case 'x':
	    set_overwrite = TRUE;
	    break;
	case 'X':
	    set_writehaof = TRUE;
	    break;
	case '0':
	    set_delete_msgnum = add_list(set_delete_msgnum, optarg);
	    break;
	case '1':
	    set_readone = TRUE;
	    break;
	case 'M':
	    set_usemeta = TRUE;
	    break;
	case 'N':
 	    set_nonsequential = TRUE;
	    break;
	case '?':
	default:
	    break;
	}
    }
 
    /* do some postconfig checks for deprecated / obsolete options and inits */
    PostConfig();
        
#ifdef DEBUG
    dump_config();
    exit(0);
#endif

    /*
     * Now override the configuration file variables with any explicitly
     * passed on the command line. This way you need not change the
     * configuration file settings for a minor change in a single run.
     */

    /* 
     * Check and make sure that the supplied language is a
     * valid language. Otherwise strange things happen quickly.
     */

    if (strlen(set_language) > 2) {
	locale_code = set_language;
	set_language[2] = 0;	/* shorten to 2-letter code */
    }
    else
	locale_code = NULL;

    if ((tlang = valid_language(set_language, &locale_code)) == NULL) {
        trio_snprintf(errmsg, sizeof(errmsg), "\"%s\" %s.", set_language, lang[MSG_LANGUAGE_NOT_SUPPORTED]);
        cmderr(errmsg);
    }

#ifdef HAVE_LOCALE_H
    if (!setlocale(LC_ALL, locale_code)) {
        char *rv = NULL;
        
        if (!strcmp(locale_code, "en_US")) {
            /* many systems now install by defualt en_US.UTF-8.
               Here we assume that the mapping between en_US and en_US.UTF-8
               in system messages is identical. 
               We cannot do the same for other languages, though */
            rv = setlocale(LC_ALL, "en_US.UTF-8");
        }
        if (!rv) {
            trio_snprintf(errmsg, sizeof(errmsg), "WARNING: locale \"%s\", not supported.\n", locale_code);
            fprintf(stderr, "%s", errmsg);/* AUDIT biege: avoid format-bug warning */
        }
    }
#endif

    lang = tlang;		/* A good language, make it so. */

    if (print_usage)		/* Print the usage message and terminate */
	usage();

#ifndef GDBM
    if (set_usegdbm) {
        fprintf(stderr, "%s: %s\n", PROGNAME, lang[MSG_OPTION_G_NOT_BUILD_IN]);
        usage();
    }
#endif
    
#ifndef HAVE_LIBFNV
    if (set_nonsequential)
        progerr("Hypermail isn't built with the libfnv hash library.\n"
                "You cannot use the nonsequential option.\n");
#endif /* HAVE_LIBFNV */
    
    if (set_mbox && !strcasecmp(set_mbox, "NONE")) {
	use_stdin = TRUE;
    }

    /* the list of headers that we always show and that we want to avoid
       showing twice when printing the body */
    set_skip_headers = add_list(set_skip_headers, "from");
    set_skip_headers = add_list(set_skip_headers, "date");
    set_skip_headers = add_list(set_skip_headers, "subject");

    /*
     * Did they decide to use stdin by specifying it on the command line ?
     * If not then check and see if that is the default they wanted from
     * the options.h or environment values.
     */
    if (!use_stdin) {
	if (optind < argc && set_increment == -1 && !set_mbox) {
	    set_mbox = strsav(argv[optind]);
	}
	else if (!set_mbox || !strcasecmp(set_mbox, "NONE"))
	    use_stdin = TRUE;
	else
	    use_stdin = FALSE;
    }
    else {
	if (set_mbox) {
	    free(set_mbox);
        }
	set_mbox = NULL;
    }

    if (set_dir) {
        char *dp = dirpath(set_dir);
	set_dir = strreplace(set_dir, dp);
        free (dp);
    }
    
    /*
     * Default names for directories and labels need to be figured out.
     */

    if (use_stdin && (!set_dir || !strcasecmp(set_dir, "NONE")))
	set_dir = strreplace(set_dir, DIRNAME);

    if (!set_dir || !strcasecmp(set_dir, "NONE"))
	set_dir = strreplace(set_dir, (strrchr(set_mbox, '/')) ? strrchr(set_mbox, '/') + 1 : set_mbox);

    if (set_dir[strlen(set_dir) - 1] != PATH_SEPARATOR) {
        char *t = set_dir;
        
        trio_asprintf(&set_dir, "%s%c", t, PATH_SEPARATOR);
        free(t);
    }

    if (!set_label || !strcasecmp(set_label, "NONE")) {
	set_label = set_mbox ? (strreplace(set_label, (strrchr(set_mbox, '/')) ? strrchr(set_mbox, '/') + 1 : set_mbox)) : "stdin";
    }
    
    /*
     * Read the contents of the file into the variables to be used
     * in printing out the pages.
     */

    ihtmlheaderfile = expand_contents(set_ihtmlheader);
    ihtmlfooterfile = expand_contents(set_ihtmlfooter);
    ihtmlheadfile = expand_contents(set_ihtmlhead);
    ihtmlhelpupfile = expand_contents(set_ihtmlhelpup);
    ihtmlhelplowfile = expand_contents(set_ihtmlhelplow);
    ihtmlnavbar2upfile = expand_contents(set_ihtmlnavbar2up);
    mhtmlheaderfile = expand_contents(set_mhtmlheader);
    mhtmlfooterfile = expand_contents(set_mhtmlfooter);
    if (set_mhtmlnavbar2up) {
        mhtmlnavbar2upfile = expand_contents(set_mhtmlnavbar2up);
    } else {
        mhtmlnavbar2upfile = expand_contents(set_ihtmlnavbar2up);
    }

    /* if the user didn't specify a message navbar for messages, we
       set up a generic one using the archives's label and linking
       back to the main index */
    if (!mhtmlnavbar2upfile || !*mhtmlnavbar2upfile) {
        trio_asprintf(&mhtmlnavbar2upfile, DEFAULT_MHTML_NAVBAR2UP, set_label);
    }
    
    /*
     * Which index file will be called "index.html"?
     */

    index_name[1][DATE_INDEX] = setindex(set_defaultindex, "date", set_htmlsuffix);
    index_name[1][THREAD_INDEX] = setindex(set_defaultindex, "thread", set_htmlsuffix);
    index_name[1][SUBJECT_INDEX] = setindex(set_defaultindex, "subject", set_htmlsuffix);
    index_name[1][AUTHOR_INDEX] = setindex(set_defaultindex, "author", set_htmlsuffix);
    if (set_attachmentsindex) {
	index_name[1][ATTACHMENT_INDEX]
	    = setindex(set_defaultindex, "attachment", set_htmlsuffix);
    }
    if (set_folder_by_date || set_msgsperfolder) {
	index_name[0][DATE_INDEX] = setindex(set_default_top_index, "date", set_htmlsuffix);
	index_name[0][THREAD_INDEX] = setindex(set_default_top_index, "thread", set_htmlsuffix);
	index_name[0][SUBJECT_INDEX] = setindex(set_default_top_index, "subject", set_htmlsuffix);
	index_name[0][AUTHOR_INDEX] = setindex(set_default_top_index, "author", set_htmlsuffix);
	if (set_attachmentsindex) {
	    index_name[0][ATTACHMENT_INDEX] = setindex(set_default_top_index, "attachment", set_htmlsuffix);
	}
	index_name[0][FOLDERS_INDEX] = setindex(set_default_top_index, "folders", set_htmlsuffix);
    }
    else {
	index_name[0][DATE_INDEX] = index_name[1][DATE_INDEX];
	index_name[0][THREAD_INDEX] = index_name[1][THREAD_INDEX];
	index_name[0][AUTHOR_INDEX] = index_name[1][AUTHOR_INDEX];
	index_name[0][SUBJECT_INDEX] = index_name[1][SUBJECT_INDEX];
	index_name[0][ATTACHMENT_INDEX] = index_name[1][ATTACHMENT_INDEX];
    }

    init_index_names();

    if (set_msgsperfolder && set_folder_by_date) {
	progerr("msgsperfolder and folder_by_date may not be used at the same time!");
    }

    /*
     * General settings for mail command and rewriting.
     */

    if (!set_domainaddr || !strcasecmp(set_domainaddr, "NONE"))
	use_domainaddr = 0;
    else
	use_domainaddr = 1;

    if (!set_mailto || !strcasecmp(set_mailto, "NONE"))
	use_mailto = 0;
    else
	use_mailto = 1;

    if (!set_mailcommand || !strcasecmp(set_mailcommand, "NONE"))
	use_mailcommand = 0;
    else
	use_mailcommand = 1;

#ifndef HAVE_LIBFNV
    /* the nonsequential mode won't work unless we compiled the FNV hash library
       (./configure --enable-libfnv) */
    if (set_nonsequential)
		progerr("the nonsequential mode is only available if you enabled the\n compilation" "of the fnv hash library. Try doing a\n\t./configure --enable-libfnv\n" "and recompile if you want to use this option.");
#endif /* HAVE_LIBFNV */

    /* 
     * A little performance speed up.  The following was being done
     * over and over in the write functions. This way it is done once.
     * A bigger win on larger archives.
     */

    if (set_hmail && !strcasecmp(set_hmail, "NONE")) {
	free(set_hmail);
	set_hmail = NULL;
    }

    if (set_archives && !strcasecmp(set_archives, "NONE")) {
	free(set_archives);
	set_archives = NULL;
    }

    if (set_custom_archives && !strcasecmp(set_custom_archives, "NONE")) {
	free(set_custom_archives);
	set_custom_archives = NULL;
    }

    if (set_about && !strcasecmp(set_about, "NONE")) {
	free(set_about);
	set_about = NULL;
    }

    /* Simply show what the values of the variables are and exit */

    if (cmd_show_variables) {
	if (!set_mbox)
	    set_mbox = "NONE";
	MakeConfig(TRUE); 
	free(configfile);
	return (0);
    }

    /* Injecting a little sanity... */

    if (use_mbox && use_stdin) {
	cmderr(lang[MSG_CANNOT_READ_FROM_BOTH_FILE_AND_STDIN]);
    }
    if (set_append && use_mbox) {
        cmderr(lang[MSG_CANNOT_BOTH_READ_AND_WRITE_TO_MBOX]);
    }

    gettimezone();
    getthisyear();

    /*
     * MIME processing requires the files be created as they
     * are read in loadheaders() so assure the directories are 
     * there first...
     */

    checkdir(set_dir);

    /* write the default css if any of the two custom ones was not
       declared */
    if (! (set_icss_url && *set_icss_url) ||
        ! (set_mcss_url && *set_mcss_url)) {
        
        char *filename;

	if (set_default_css_url && !strcmp (set_default_css_url, "hypermail.css")) {
            trio_asprintf(&filename, "%s%s", set_dir, "hypermail.css");
            if (!isfile(filename)) {
                print_default_css_file (filename);
            }
            free(filename);
	}
    }
	
    /*
     * Let's do it.
     */

    if (set_uselock)
	lock_archive(set_dir);

    if (set_increment == -1) {
	int save_append = set_append;
	set_append = 0;
	if (set_mbox_shortened)
	    progerr("can not use increment = -1 option with mbox_shortened option\n");
	amount_new = parsemail(set_mbox, use_stdin, 1, -1, set_dir, set_inlinehtml, 0);
	set_increment = !matches_existing(set_startmsgnum);
	if (set_increment && set_folder_by_date && !set_usegdbm)
	    progerr("folder_by_date with incremental update requires usegdbm option");
	reinit_structs();
	set_append = save_append;
    }
    if (set_increment) {
	int num_displayable;
	int num_added;
	if (set_linkquotes)
	    replylist = NULL;
	/* we have to start with the msgnum - 1 so that the rest of the
	   code works ok when there are no old headers. */
	max_msgnum = set_startmsgnum - 1;
	num_displayable = loadoldheaders(set_dir);
	amount_old = max_msgnum + 1; /* counts gaps as messages */

	/* start numbering at this number */
	num_added = parsemail(set_mbox, use_stdin, set_readone, set_increment, set_dir, set_inlinehtml, amount_old);
	if (num_added > 0) {
	    amount_new = num_displayable + num_added;
	    if (set_linkquotes)
		analyze_headers(max_msgnum + 1);

	    /* write the index of msgno/msgid_hash filenames */
	    if (set_nonsequential)
		    write_messageindex(0, max_msgnum + 1);

	    writearticles(amount_old, max_msgnum + 1);

	    /* JK: in function of other hypermail configuration options, 
	       delete_incremental will continuous escape and add more markup
	       to non-deleted messages that are replies to deleted messages.
	       Thus, a setup option to disable it */
	    if (set_delete_incremental && deletedlist)
		update_deletions(amount_old);

	    if (set_show_msg_links) {
		fixnextheader(set_dir, amount_old, -1);
		for (i = amount_old; i <= max_msgnum; ++i) {
		    if (set_showreplies)
			fixreplyheader(set_dir, i, 0, amount_old);
		    fixthreadheader(set_dir, i, amount_old);
		}
	    }
	}
    }
    else {
	if (set_mbox_shortened) {
	    if (!set_usegdbm) progerr("mbox_shortened option requires that the usegdbm option be on");
	    max_msgnum = set_startmsgnum - 1;
	    loadoldheaders(set_dir);
	}
	amount_new = parsemail(set_mbox, use_stdin, set_readone, set_increment, set_dir, 
			       set_inlinehtml, set_startmsgnum);	/* number from 0 */
	if (!set_mbox_shortened && !matches_existing(0)) {
	    progerr("First message in mailbox does not "
		    "match first message in archive\n"
		    "or obsolete gdbm file present.\n"
		    "Maybe you want to enable the mbox_shortened option?\n");
	}
	if (set_linkquotes)
	    analyze_headers(max_msgnum + 1);

	/* write the index of msgno/msgid_hash filenames */
	if (set_nonsequential)
		write_messageindex(0, max_msgnum + 1);

	writearticles(0, max_msgnum + 1);
    }

    /* only update the indices when we either have a new
       message (if set_increment) is set or when inputting
       a whole mbox. 
       This is a lame way to take into account msgid collision
       when adding a new message. Without binding the condition
       to set_condition, this could result in the indices stating
       the archive is empty */
    if ((set_increment && amount_new > 0)
        || (!set_increment
            && (amount_new > 0 || count_deleted(max_msgnum + 1)))) {
	/* Always write the index files */
	if (amount_new && set_linkquotes) {
	    threadlist = NULL;
	    threadlist_end = NULL;
	    printedthreadlist = NULL;
	    for (i = 0; i <= max_msgnum; ++i) {
	        struct emailinfo *ep;
		if (hashnumlookup(i, &ep)) {
		    ep->flags &= ~USED_THREAD;
#ifdef FASTREPLYCODE
		    ep->isreply = 0;
#endif
		}
		threadlist_by_msgnum[i] = NULL;
	    } /* redo threading with more complete info than in 1st pass */
	    crossindexthread1(datelist);
	    for (i = 0; i <= max_msgnum; ++i) {
	        struct emailinfo *ep, *etmp;
		hashnumlookup(i, &ep);
		etmp = nextinthread(i);
		if (etmp && ep->initial_next_in_thread != etmp->msgnum)
		    fixthreadheader(set_dir, etmp->msgnum, amount_new);
		/* if (ep->flags & THREADING_ALTERED) */
	    }
	}
        
	if (show_index[0][DATE_INDEX])
	    writedates(amount_new, NULL);
	if (show_index[0][THREAD_INDEX])
	    writethreads(amount_new, NULL);
	if (show_index[0][SUBJECT_INDEX])
	    writesubjects(amount_new, NULL);
	if (show_index[0][AUTHOR_INDEX])
	    writeauthors(amount_new, NULL);
	if (set_attachmentsindex) {
	    writeattachments(amount_new, NULL);
	}
	if (set_writehaof) 
            writehaof(amount_new, NULL);
	if (set_folder_by_date || set_msgsperfolder)
	    write_toplevel_indices(amount_new);
	if (set_monthly_index || set_yearly_index)
	    write_summary_indices(amount_new);
	if (set_latest_folder)
	    symlink_latest();
    }
    else {
	printf("No mails to output!\n");
    }

    if (set_uselock)
	unlock_archive();

    /* 
    ** do some cleanup 
    */
    printed_free(printedlist);
    printed_free(printedthreadlist);

    ConfigCleanup();
    
    if (configfile)
	free(configfile);
    
    return (0);
}
